mesonpy/__init__.py

```diff
@@ -972,12 +1011,16 @@ def get_requires_for_build_wheel(
     config_settings: Optional[Dict[str, str]] = None,
 ) -> List[str]:
     dependencies = [_depstr.wheel]
-    with _project(config_settings) as project:
-        if not project.is_pure and platform.system() == 'Linux':
-            # we may need patchelf
-            if not shutil.which('patchelf'):  # XXX: This is slightly dangerous.
-                # patchelf not already acessible on the system
-                dependencies.append(_depstr.patchelf)
+
+    if _env_ninja_command() is None:
+        dependencies.append(_depstr.ninja)
+
+    if sys.platform.startswith('linux'):
+        # we may need patchelf
+        if not shutil.which('patchelf'):
+            # patchelf not already accessible on the system
+            dependencies.append(_depstr.patchelf)
+
```

Comment:
Please do not change the `patchelf` logic here, this is a regression of #158.

Reply:
I can check at home, but I'm pretty sure you can't use `_project` which uses `meson` without `ninja`. I think the only loss is for pure python projects, which probably shouldn't be using meson?

Reply:
You should be able to, Meson generates the introspection metadata during the build directory setup. ninja should only be needed for the build step, which only happens when `build_wheel` is called.

Reply:
@henryiii the problem is about `patchelf`, not `ninja`. You lost a `if not project.is_pure` in the condition.

Reply:
> You should be able to, Meson generates the introspection metadata during the build directory setup. ninja should only be needed for the build step, which only happens when `build_wheel` is called.

AFAIK, `meson setup` fails if it does not find `ninja`.

Reply:
> I think the only loss is for pure python projects, which probably shouldn't be using meson?

Not actually true, tbh. There's a couple reasons why you might want to use Meson, for example, Meson wants to use Meson. :D Also, if you have files that need to be installed outside of a python install scheme, for example configuration files to /etc or, when installing as a user, config files to `~/.config/foobar.ini`, scripts to `~/.local/bin/foobar`, desktop files to `~/.local/share/applications/foobar.desktop`, and python packages to `~/.venvs/foobar/lib/python3.X/site-packages/foobar-1.0.dist-info/`.

Admittedly, that use case is "do not use wheels", so it would be an enhanced mode when using Meson directly instead of via this backend, but the principle is there. :D 

I also have a Meson PR to add `meson setup builddir/ --backend=none` that doesn't generate a ninja backend, only installdata, which could be useful for pure python projects (again, such as Meson itself).

Reply:
We can't run meson inside get_requires to find out if a project is pure unless meson can be run without ninja, I believe, without a "backend=none" feature. The only loss here is a pure Python project will require patchelf (possibly pre-installed) on linux systems until such a feature is added to meson. Which seems a lot better the current ninja requirement, which is hurting both pure and non pure projects.

Reply:
Okay, then we should do the test regardless if ninja is available, odds are if you are using a local patchelf executable, you are also using a local ninja, and always add patchelf if not. Hopefully, we will be able to change that soon, if/when Eli's PR gets in.

Reply:
Then please express your support for https://github.com/mesonbuild/meson/pull/10779 :)

Reply:
@FFY00 Is there a change necessary? Also, this could be solved by adding a TOML option turning off patchelf checking. Something like `tool.meson-python.needs-patchelf=False` could disable the check entirely, and if you have a pure Python package, you could disable the check (also useful if you don't need patchelf for other reasons).  This avoids paying the cost of running meson just to get need_requires, too.

Reply:
One alternate suggestion: For packages that have a known set of platforms, you can check to see if the platform is supported by a binary wheel, and only add it if it is. See, for example, https://github.com/maxbachmann/RapidFuzz/blob/e6bc56c1adf18b19008c99be6e3318b8d895c10f/_custom_build/backend.py#LL33. We could do that for patchelf, which would only add it if it's safe to add (supported by a wheel), which would allow pure Python packages to avoid breaking if on a system with no patchelf and no wheel.

Reply:
I think there is nothing to change in this patch. I'm working on adding some tests, but testing this from the perspective of running `python -m build ...` makes the tests very ugly, especially trying to avoid repeating expensive test setup...

Reply:
I pushed a commit with the changes I was asking. Please review and let me know if you think there will be any issues.

Reply:
Okay, that's fine. I'm still not sure forcing all projects to run `meson setup` just so you can keep pure Python projects from failing (in an environment where binary projects will still fail, mind you!) is a necessary optimization to make. Regardless, hopefully https://github.com/mesonbuild/meson/pull/10779 fixes the no-ninja case or there's eventually a patchelf configuration option, and it won't matter, and it's not any slower or worse than what is done before this PR, it's clearly a net improvement.

Reply:
(I underwent what you meant now :) )